Fallback bootstrap router build parameter, ignore obsolete bootstrap routers

tewinget opened the pull request:
Addresses #1995 .

jagerman committed:
Significantly decrease decomm warning frequency

30s is insanely fast for a error level log statement and spams logs
incessantly; reduce it to repeating once every 5 minutes.

jagerman committed:
Clean up/simplify SN list handling

tewinget committed:
Fallback bootstrap router build parameter

Adds a fallback bootstrap file path parameter to CMake, specify
-DBOOTSTRAP_SYSTEM_PATH="/path/to/file" to use.

Adds a list of (currently 1) obsolete bootstrap RouterIDs to check
bootstrap RCs against.  Will not use bootstrap RCs if they're on that
list.

Log an error periodically if we appear to be an active service node but
have fewer than a set number (5) known peers.

Bumps oxen-logging version for literal _format.

jagerman commented on llarp/router_contact.cpp:
```diff
@@ -532,6 +532,21 @@ namespace llarp
+  bool
+  RouterContact::IsObsoleteBootstrap() const
+  {
+    constexpr std::array obsolete_bootstraps = {
```
Should move this outside of the function (and make it static constexpr)

jagerman commented on llarp/router/router.cpp:
```diff
@@ -666,39 +666,24 @@ namespace llarp
+        rc_itr++;
+    }
+
     for (auto& rc : b_list)
```
This for loop can be DRYed out a bit with the nearly identical loop below

jagerman commented on llarp/router/router.cpp:
```diff
@@ -711,6 +696,24 @@ namespace llarp
 
     if (bootstrapRCList.empty() and not conf.bootstrap.seednode)
     {
+#ifdef BOOTSTRAP_FALLBACK
```
I would rather keep the `ifdef` as small as possible to minimize code compilation differences, perhaps via this outside the function:
```C++
static constexpr std::optional<std::string_view> bootstrap_fallback_path{
#ifdef BOOTSTRAP_FALLBACK
    std::string_view{BOOTSTRAP_FALLBACK}
#else
    std::nullopt;
#endif
};
```
Then the code here can do (without an ifdef):
```C++
if (bootstrap_fallback_path)
{
  ...
}
```

jagerman commented on llarp/router/router.cpp:
```diff
@@ -666,52 +666,57 @@ namespace llarp
+        b_list.clear();
+        b_list.AddFromFile(BOOTSTRAP_FALLBACK);
+
+        verifyRCs(false);
```
```suggestion
        verifyRCs(true);
```

jagerman commented on llarp/router/router.cpp:
```diff
@@ -666,52 +666,57 @@ namespace llarp
+          LogWarn("ignoring invalid RC: ", RouterID(rc.pubkey));
+          continue;
+        }
+        if (ignoreObsolete and rc.IsObsoleteBootstrap())
```
Should ~~LogWarn~~ `log::warning(logcat, ...)` here as well to let the admin know there is something being skipped in the bootstrap file.

jagerman commented on llarp/router/router.cpp:
```diff
@@ -1032,14 +1037,27 @@ namespace llarp
+      {
+        constexpr int KnownPeerWarningThreshold = 5;
+        if (nodedb()->NumLoaded() < KnownPeerWarningThreshold)
+          LogError(
+              "We appear to be an active service node, but have fewer than ",
+              KnownPeerWarningThreshold,
+              " known peers.");
```
```suggestion
          log::error(logcat,
              "We appear to be an active service node, but have fewer than {} known peers", KnownPeerWarningThreshold);
```

jagerman commented on llarp/router/router.cpp:
```diff
@@ -666,52 +666,60 @@ namespace llarp
-        continue;
+        if (rc.IsObsoleteBootstrap())
+        {
+          LogWarn("ignoring obsolete boostrap RC: ", RouterID(rc.pubkey));
```
```suggestion
          log::warning(logcat, "ignoring obsolete boostrap RC: {}", RouterID(rc.pubkey));
```

majestrate commented on llarp/bootstrap.cpp:
```diff
@@ -36,4 +36,35 @@ namespace llarp
+        throw std::runtime_error{
+            fmt::format("failed to decode bootstrap RC, file='{}', rc={}", fpath, rc)};
+      }
+      this->insert(rc);
```
since you're here, we really need to not make BootstrapList a subtype of std::vector

jagerman replied:
... the fuck?

majestrate replied:
my bad. std::set

jagerman replied:
Should be done, but seems a bit out of scope for this PR.

tewinget replied:
std::set, and while that is something that would be good it's out of scope for this PR

jagerman commented:
One more thing:

Let's make lokinet-router *not* ping oxend if it's active and doesn't know of at least (threshold) number of routers, because that generally indicates some sort of serious connectivity issue and we want the operator to notice right away (because of a lack of proofs).  (We *do* still want it to ping oxend while unfunded, though, because it's common for an unfunded router to not know about any routers).

tewinget committed:
don't ping core if active with too few peers

If running as a service node, we ping core on a regular interval to
inform it we're running and in a good state.  If we're an active
(not decommissioned or deregistered) service node and have too few
peers and thus we're not actually connected to lokinet, we should skip
that ping so core doesn't think we're ok.

jagerman reviewed (changes requested):
Two changes needed here:

- the "fallback" should just bake the bootstrap data into the binary at compile time rather than carrying a path around that we have to distribute.
- we need to support bootstraps for different networks, typically mainnet + testnet, so that compile time fallback should probably be a unordered_map of netid -> bootstrap data.

jagerman committed:
bake in bootstrap fallbacks at build time

also change the defailt/fallback bootstrap file for testnet to be a list
of RCs (albeit of size 1) rather than just an RC dict.

jagerman committed:
oxen-mq update to tagged stable release

jagerman committed:
Fix obsolete oxenmq/hex.h usage
